### ISSUE TITLE
CI: Multiple tests: Timed out waiting for datapath updates of FQDN IP information; returning response

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#22432]
### Test Name

```shell
K8sAgentHubbleTest Hubble Observe Test FQDN Policy with Relay
K8sAgentPolicyTest Basic Test TLS policy
K8sAgentPolicyTest Basic Test Traffic redirections to proxy Tests DNS proxy visibility without policy
```

### Failure Output

```shell
Found 1 k8s-app=cilium logs matching list of errors that must be investigated:
level=error
```

### Stack Trace

```shell
/home/jenkins/workspace/Cilium-PR-K8s-1.16-kernel-4.9/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:415
Found 1 k8s-app=cilium logs matching list of errors that must be investigated:
level=error
/home/jenkins/workspace/Cilium-PR-K8s-1.16-kernel-4.9/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:413
```

### Standard Output

```shell
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 0
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
No errors/warnings found in logs
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 0
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
No errors/warnings found in logs
⚠️  Found "level=error" in logs 1 times
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 1
Number of "level=warning" in logs: 0
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
Top 1 errors/warnings:
Timed out waiting for datapath updates of FQDN IP information; returning response
Cilium pods: [cilium-j9znj cilium-w5v9b]
Netpols loaded: 
CiliumNetworkPolicies loaded: 
Endpoint Policy Enforcement:
Pod                            Ingress   Egress
app3-6c7856c5b5-mrb2d          false     false
coredns-8cfc78c54-9lfqg        false     false
hubble-relay-cb7fbcc49-wqsgt   false     false
app1-7b6ddb776f-5sg6d          false     false
app1-7b6ddb776f-f6mhf          false     false
app2-5cc5d58844-7qx2x          false     false
Cilium agent 'cilium-j9znj': Status: Ok  Health: Ok Nodes "" ContainerRuntime:  Kubernetes: Ok KVstore: Ok Controllers: Total 35 Failed 0
Cilium agent 'cilium-w5v9b': Status: Ok  Health: Ok Nodes "" ContainerRuntime:  Kubernetes: Ok KVstore: Ok Controllers: Total 29 Failed 0
```

### Standard Error

```shell
=== Test Finished at 2022-11-29T17:29:34Z====
17:29:34 STEP: Running JustAfterEach block for EntireTestsuite K8sAgentHubbleTest Hubble Observe
FAIL: Found 1 k8s-app=cilium logs matching list of errors that must be investigated:
level=error
===================== TEST FAILED =====================
17:29:34 STEP: Running AfterFailed block for EntireTestsuite K8sAgentHubbleTest Hubble Observe
cmd: kubectl get pods -o wide --all-namespaces
Exitcode: 0 
Stdout:
 	 NAMESPACE                                                 NAME                            READY   STATUS    RESTARTS   AGE     IP              NODE   NOMINATED NODE   READINESS GATES
	 202211291729k8sagenthubbletesthubbleobservetestl3l4flow   app1-7b6ddb776f-5sg6d           2/2     Running   0          19s     10.0.1.125      k8s1   <none>           <none>
	 202211291729k8sagenthubbletesthubbleobservetestl3l4flow   app1-7b6ddb776f-f6mhf           2/2     Running   0          19s     10.0.1.137      k8s1   <none>           <none>
	 202211291729k8sagenthubbletesthubbleobservetestl3l4flow   app2-5cc5d58844-7qx2x           1/1     Running   0          19s     10.0.1.20       k8s1   <none>           <none>
	 202211291729k8sagenthubbletesthubbleobservetestl3l4flow   app3-6c7856c5b5-mrb2d           1/1     Running   0          19s     10.0.1.205      k8s1   <none>           <none>
	 cilium-monitoring                                         grafana-7fd557d749-gdxmz        0/1     Running   0          32m     10.0.1.174      k8s2   <none>           <none>
	 cilium-monitoring                                         prometheus-d87f8f984-x2hvx      1/1     Running   0          32m     10.0.1.77       k8s2   <none>           <none>
	 kube-system                                               cilium-j9znj                    1/1     Running   0          97s     192.168.56.11   k8s1   <none>           <none>
	 kube-system                                               cilium-operator-7675967-cx9w5   1/1     Running   0          97s     192.168.56.11   k8s1   <none>           <none>
	 kube-system                                               cilium-operator-7675967-g5szd   1/1     Running   0          97s     192.168.56.12   k8s2   <none>           <none>
	 kube-system                                               cilium-w5v9b                    1/1     Running   0          97s     192.168.56.12   k8s2   <none>           <none>
	 kube-system                                               coredns-8cfc78c54-9lfqg         1/1     Running   0          2m32s   10.0.0.20       k8s2   <none>           <none>
	 kube-system                                               etcd-k8s1                       1/1     Running   0          35m     192.168.56.11   k8s1   <none>           <none>
	 kube-system                                               hubble-relay-cb7fbcc49-wqsgt    1/1     Running   0          97s     10.0.0.120      k8s2   <none>           <none>
	 kube-system                                               kube-apiserver-k8s1             1/1     Running   0          35m     192.168.56.11   k8s1   <none>           <none>
	 kube-system                                               kube-controller-manager-k8s1    1/1     Running   2          35m     192.168.56.11   k8s1   <none>           <none>
	 kube-system                                               kube-proxy-25qlb                1/1     Running   0          36m     192.168.56.11   k8s1   <none>           <none>
	 kube-system                                               kube-proxy-ftvkp                1/1     Running   0          33m     192.168.56.12   k8s2   <none>           <none>
	 kube-system                                               kube-scheduler-k8s1             1/1     Running   2          35m     192.168.56.11   k8s1   <none>           <none>
	 kube-system                                               log-gatherer-s2vx4              1/1     Running   0          32m     192.168.56.12   k8s2   <none>           <none>
	 kube-system                                               log-gatherer-vwlqs              1/1     Running   0          32m     192.168.56.11   k8s1   <none>           <none>
	 kube-system                                               registry-adder-5gv7t            1/1     Running   0          33m     192.168.56.12   k8s2   <none>           <none>
	 kube-system                                               registry-adder-wdnhz            1/1     Running   0          33m     192.168.56.11   k8s1   <none>           <none>
	 
Stderr:
 	 

Fetching command output from pods [cilium-j9znj cilium-w5v9b]
cmd: kubectl exec -n kube-system cilium-j9znj -c cilium-agent -- cilium endpoint list
Exitcode: 0 
Stdout:
 	 ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                                                               IPv6        IPv4         STATUS   
	            ENFORCEMENT        ENFORCEMENT                                                                                                                                     
	 1207       Disabled           Disabled          15575      k8s:id=app3                                                                               fd02::14a   10.0.1.205   ready   
	                                                            k8s:io.cilium.k8s.policy.cluster=default                                                                                   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=default                                                                            
	                                                            k8s:io.kubernetes.pod.namespace=202211291729k8sagenthubbletesthubbleobservetestl3l4flow                                    
	                                                            k8s:zgroup=testapp                                                                                                         
	 1377       Disabled           Disabled          4          reserved:health                                                                           fd02::15e   10.0.1.203   ready   
	 1693       Disabled           Disabled          29278      k8s:id=app1                                                                               fd02::198   10.0.1.137   ready   
	                                                            k8s:io.cilium.k8s.policy.cluster=default                                                                                   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=app1-account                                                                       
	                                                            k8s:io.kubernetes.pod.namespace=202211291729k8sagenthubbletesthubbleobservetestl3l4flow                                    
	                                                            k8s:zgroup=testapp                                                                                                         
	 2031       Disabled           Disabled          21579      k8s:appSecond=true                                                                        fd02::1f2   10.0.1.20    ready   
	                                                            k8s:id=app2                                                                                                                
	                                                            k8s:io.cilium.k8s.policy.cluster=default                                                                                   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=app2-account                                                                       
	                                                            k8s:io.kubernetes.pod.namespace=202211291729k8sagenthubbletesthubbleobservetestl3l4flow                                    
	                                                            k8s:zgroup=testapp                                                                                                         
	 2061       Disabled           Disabled          29278      k8s:id=app1                                                                               fd02::153   10.0.1.125   ready   
	                                                            k8s:io.cilium.k8s.policy.cluster=default                                                                                   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=app1-account                                                                       
	                                                            k8s:io.kubernetes.pod.namespace=202211291729k8sagenthubbletesthubbleobservetestl3l4flow                                    
	                                                            k8s:zgroup=testapp                                                                                                         
	 2548       Disabled           Disabled          1          k8s:cilium.io/ci-node=k8s1                                                                                         ready   
	                                                            k8s:node-role.kubernetes.io/master                                                                                         
	                                                            reserved:host                                                                                                              
	 
Stderr:
 	 

cmd: kubectl exec -n kube-system cilium-w5v9b -c cilium-agent -- cilium endpoint list
Exitcode: 0 
Stdout:
 	 ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                            IPv6       IPv4         STATUS   
	            ENFORCEMENT        ENFORCEMENT                                                                                                 
	 544        Disabled           Disabled          5827       k8s:app.kubernetes.io/name=hubble-relay                fd02::1a   10.0.0.120   ready   
	                                                            k8s:app.kubernetes.io/part-of=cilium                                                   
	                                                            k8s:io.cilium.k8s.policy.cluster=default                                               
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=hubble-relay                                   
	                                                            k8s:io.kubernetes.pod.namespace=kube-system                                            
	                                                            k8s:k8s-app=hubble-relay                                                               
	 579        Disabled           Disabled          4          reserved:health                                        fd02::9c   10.0.0.88    ready   
	 1013       Disabled           Disabled          40749      k8s:io.cilium.k8s.policy.cluster=default               fd02::c7   10.0.0.20    ready   
	                                                            k8s:io.cilium.k8s.policy.serviceaccount=coredns                                        
	                                                            k8s:io.kubernetes.pod.namespace=kube-system                                            
	                                                            k8s:k8s-app=kube-dns                                                                   
	 3696       Disabled           Disabled          1          k8s:cilium.io/ci-node=k8s2                                                     ready   
	                                                            reserved:host                                                                          
	 
Stderr:
 	 

===================== Exiting AfterFailed =====================
17:29:43 STEP: Running AfterEach for block EntireTestsuite

[[ATTACHMENT|8ca74a84_K8sAgentHubbleTest_Hubble_Observe_Test_FQDN_Policy_with_Relay.zip]]
17:29:44 STEP: Running AfterAll block for EntireTestsuite K8sAgentHubbleTest Hubble Observe
17:29:44 STEP: Deleting namespace 202211291729k8sagenthubbletesthubbleobservetestl3l4flow
17:30:00 STEP: DeleteHubbleRelay(namespace="kube-system")
17:30:12 STEP: Removing Cilium installation using generated helm manifest
```

### Resources

- Jenkins URL: https://jenkins.cilium.io/job/cilium-master-k8s-1.21-kernel-4.9/369/testReport/junit/Suite-k8s-1/21/K8sAgentHubbleTest_Hubble_Observe_Test_FQDN_Policy_with_Relay/
- ZIP file(s): 
[9d6a3778_K8sAgentHubbleTest_Hubble_Observe_Test_FQDN_Policy_with_Relay.zip](https://github.com/cilium/cilium/files/10118013/9d6a3778_K8sAgentHubbleTest_Hubble_Observe_Test_FQDN_Policy_with_Relay.zip)

**Comment by joestringer** [comment]
Some relevant logs from `pod-kube-system-cilium-qpc95-cilium-agent.log`, a few logs before + after the error:

```
2022-11-29T05:45:16.255657611Z level=debug msg="Forwarding DNS request for a name that is allowed" DNSRequestID=34520 dnsName=vagrant-cache.ci.cilium.io. endpointID=1086 identity=17132 ipAddr="10.0.0.239:48023" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.255659868Z level=debug msg="Logging flow record" message="Allowed by policy" subsys=proxy-logger type=Request verdict=Forwarded
2022-11-29T05:45:16.255854872Z level=debug msg="Handling DNS query from endpoint" DNSRequestID=34666 dnsName=vagrant-cache.ci.cilium.io. ipAddr="10.0.0.239:48023" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.255860541Z level=debug msg="Found target server to of DNS request secID {ID:11615 Source:custom-resource shadowed:false}" DNSRequestID=34666 dnsName=vagrant-cache.ci.cilium.io. endpointID=1086 identity=17132 ipAddr="10.0.0.239:48023" server="10.0.1.98:53" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.255885856Z level=debug msg="Forwarding DNS request for a name that is allowed" DNSRequestID=34666 dnsName=vagrant-cache.ci.cilium.io. endpointID=1086 identity=17132 ipAddr="10.0.0.239:48023" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.255889918Z level=debug msg="Logging flow record" message="Allowed by policy" subsys=proxy-logger type=Request verdict=Forwarded
2022-11-29T05:45:16.255891994Z level=debug msg="stale identity observed" identity=17132 ipAddr=10.0.0.26 oldIdentity=1 subsys=hubble
2022-11-29T05:45:16.255916689Z level=debug msg="stale identity observed" identity=17132 ipAddr=10.0.0.26 oldIdentity=1 subsys=hubble
2022-11-29T05:45:16.256089286Z level=debug msg="stale identity observed" identity=17132 ipAddr=10.0.0.26 oldIdentity=1 subsys=hubble
2022-11-29T05:45:16.256116365Z level=debug msg="stale identity observed" identity=17132 ipAddr=10.0.0.26 oldIdentity=1 subsys=hubble
2022-11-29T05:45:16.263944684Z level=debug msg="Received DNS response to proxied lookup" DNSRequestID=34666 dnsName=vagrant-cache.ci.cilium.io. endpointID=1086 identity=17132 ipAddr="10.0.0.239:48023" resp=";; opcode: QUERY, status: NOERROR, id: 63754\n;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 0\n\n;; QUESTION SECTION:\n;vagrant-cache.ci.cilium.io.\tIN\t AAAA\n\n;; AUTHORITY SECTION:\ncilium.io.\t40\tIN\tSOA\tns1.dnsimple.com. admin.dnsimple.com. 1666134161 86400 7200 604800 300\n" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.263963968Z level=debug msg="Notifying with DNS response to original DNS query" DNSRequestID=34666 dnsName=vagrant-cache.ci.cilium.io. endpointID=1086 identity=17132 ipAddr="10.0.0.239:48023" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.264003887Z level=debug msg="Logging flow record" message="Allowed by policy" subsys=proxy-logger type=Response verdict=Forwarded
2022-11-29T05:45:16.264008572Z level=debug msg="Responding to original DNS query" DNSRequestID=34666 dnsName=vagrant-cache.ci.cilium.io. endpointID=1086 identity=17132 ipAddr="10.0.0.239:48023" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.264010702Z level=debug msg="dnsproxy: Wrote DNS response (123/123 bytes) from 10.0.1.98:53 to 10.0.0.239:48023" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.352152158Z level=debug msg="Received DNS response to proxied lookup" DNSRequestID=34520 dnsName=vagrant-cache.ci.cilium.io. endpointID=1086 identity=17132 ipAddr="10.0.0.239:48023" resp=";; opcode: QUERY, status: NOERROR, id: 64093\n;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0\n\n;; QUESTION SECTION:\n;vagrant-cache.ci.cilium.io.\tIN\t A\n\n;; ANSWER SECTION:\nvagrant-cache.ci.cilium.io.\t60\tIN\tA\t147.75.38.95\n" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.352179684Z level=debug msg="Notifying with DNS response to original DNS query" DNSRequestID=34520 dnsName=vagrant-cache.ci.cilium.io. endpointID=1086 identity=17132 ipAddr="10.0.0.239:48023" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.352182663Z level=debug msg="Logging flow record" message="Allowed by policy" subsys=proxy-logger type=Response verdict=Forwarded
2022-11-29T05:45:16.352188030Z level=debug msg="Recording DNS lookup in endpoint specific cache" endpointID=1086 subsys=daemon
2022-11-29T05:45:16.352190163Z level=debug msg="Updating DNS name in cache from response to query" ips="[147.75.38.95]" qname=vagrant-cache.ci.cilium.io. subsys=daemon
2022-11-29T05:45:16.352244846Z level=debug msg="Updated FQDN with new IPs" IPs="[147.75.38.95]" fqdnSelectorsToUpdate="map[{ vagrant-cache.ci.cilium.io}:{}]" matchName=vagrant-cache.ci.cilium.io. subsys=fqdn
2022-11-29T05:45:16.352254075Z level=debug msg=mapSelectorsToIPs fqdnSelectors="map[{ vagrant-cache.ci.cilium.io}:{}]" subsys=fqdn
2022-11-29T05:45:16.352256563Z level=debug msg="Emitting matching DNS Name -> IPs for FQDNSelector" DNSName=vagrant-cache.ci.cilium.io. IPs="[147.75.38.95]" matchPattern=vagrant-cache.ci.cilium.io subsys=fqdn
2022-11-29T05:45:16.352288600Z level=debug msg="getting identities for IPs associated with FQDNSelector" fqdnSelector="{ vagrant-cache.ci.cilium.io}" ips="[147.75.38.95]" subsys=daemon
2022-11-29T05:45:16.352303900Z level=debug msg="Resolving identity" identityLabels="cidr:147.75.38.95/32,reserved:world" subsys=identity-cache
2022-11-29T05:45:16.352307941Z level=debug msg="updating FQDN selector" fqdnSelectorString="{ vagrant-cache.ci.cilium.io}" identitySlice="[16777217]" subsys=daemon
2022-11-29T05:45:16.352324438Z level=error msg="Timed out waiting for datapath updates of FQDN IP information; returning response" subsys=daemon
2022-11-29T05:45:16.352375131Z level=debug msg="Waited for endpoints to regenerate due to a DNS response" duration="146.383µs" endpointID=1086 qname=vagrant-cache.ci.cilium.io. subsys=daemon
2022-11-29T05:45:16.352387029Z level=debug msg="Upserting IP into ipcache layer" identity="{16777217 generated false}" ipAddr=147.75.38.95/32 key=0 subsys=ipcache
2022-11-29T05:45:16.352389224Z level=debug msg="Daemon notified of IP-Identity cache state change" identity="{16777217 generated false}" ipAddr="{147.75.38.95 ffffffff}" modification=Upsert subsys=datapath-ipcache
2022-11-29T05:45:16.352489856Z level=debug msg="1 resource names requested, filtering resources" subsys=xds xdsAckedVersion=0 xdsClientNode= xdsTypeURL=type.googleapis.com/cilium.NetworkPolicyHosts
2022-11-29T05:45:16.352495401Z level=debug msg="resource not found" subsys=xds xdsAckedVersion=0 xdsClientNode= xdsResourceName=16777217 xdsTypeURL=type.googleapis.com/cilium.NetworkPolicyHosts
2022-11-29T05:45:16.352497465Z level=debug msg="returning 0 resources out of 1 requested" subsys=xds xdsAckedVersion=0 xdsClientNode= xdsTypeURL=type.googleapis.com/cilium.NetworkPolicyHosts
2022-11-29T05:45:16.352499353Z level=debug msg="preparing new cache transaction: upserting 1 entries, deleting 0 entries" subsys=xds xdsCachedVersion=33 xdsTypeURL=type.googleapis.com/cilium.NetworkPolicyHosts
2022-11-29T05:45:16.352501334Z level=debug msg="inserting resource into cache" subsys=xds xdsCachedVersion=33 xdsResourceName=16777217 xdsTypeURL=type.googleapis.com/cilium.NetworkPolicyHosts
2022-11-29T05:45:16.352503242Z level=debug msg="committing cache transaction and notifying of new version" subsys=xds xdsCachedVersion=33 xdsTypeURL=type.googleapis.com/cilium.NetworkPolicyHosts
2022-11-29T05:45:16.352505172Z level=debug msg="Responding to original DNS query" DNSRequestID=34520 dnsName=vagrant-cache.ci.cilium.io. endpointID=1086 identity=17132 ipAddr="10.0.0.239:48023" subsys=fqdn/dnsproxy
2022-11-29T05:45:16.352535229Z level=debug msg="dnsproxy: Wrote DNS response (86/86 bytes) from 10.0.1.98:53 to 10.0.0.239:48023" subsys=fqdn/dnsproxy
```

One of the things I struggle to understand here is which DNS request / response is associated with the message `Timed out waiting for datapath updates of FQDN IP information; returning response`, and hence which request has trouble + what the corresponding messages are for that event. At a minimum, maybe the next step on this issue is just to add an event UUID for DNS requests+responses that are handled in the proxy.

**Comment by joestringer** [comment]
Possibly caused by https://github.com/cilium/cilium/pull/22346 , fixed by #22426. @jrajahalme does that sound correct to you?

**Comment by joestringer** [comment]
I think that this is a most plausible option for this is that #22426 fixed it, since CI was reliably failing on master net-next jobs for about 8 jobs in a row, but now it's been passing for about the same number since that fix was merged yesterday. https://jenkins.cilium.io/job/cilium-master-k8s-1.26-kernel-net-next/
